bus-mapping/src/evm/opcodes.rs

```diff
@@ -350,307 +363,19 @@ pub fn gen_associated_ops(
-                ),
-                (
-                    CallContextField::CallDataLength,
-                    state.tx.input.len().into(),
-                ),
-                (CallContextField::Value, call.value),
-                (CallContextField::IsStatic, (call.is_static as usize).into()),
-                (CallContextField::LastCalleeId, 0.into()),
-                (CallContextField::LastCalleeReturnDataOffset, 0.into()),
-                (CallContextField::LastCalleeReturnDataLength, 0.into()),
-                (CallContextField::IsRoot, 1.into()),
-                (CallContextField::IsCreate, 1.into()),
-                (CallContextField::CodeHash, call.code_hash.to_word()),
-            ] {
-                state.call_context_write(&mut exec_step, call.call_id, field, value);
-            }
-            Ok(exec_step)
-        }
-        // 2. Call to precompiled.
-        (_, true, _) => {
-            evm_unimplemented!("Call to precompiled is left unimplemented");
-            Ok(exec_step)
-        }
-        (_, _, is_empty_code_hash) => {
-            // 3. Call to account with empty code.
-            if is_empty_code_hash {
-                return Ok(exec_step);
-            }
-
-            // 4. Call to account with non-empty code.
-            for (field, value) in [
-                (CallContextField::Depth, call.depth.into()),
-                (
-                    CallContextField::CallerAddress,
-                    call.caller_address.to_word(),
-                ),
-                (CallContextField::CalleeAddress, call.address.to_word()),
-                (
-                    CallContextField::CallDataOffset,
-                    call.call_data_offset.into(),
-                ),
-                (
-                    CallContextField::CallDataLength,
-                    call.call_data_length.into(),
-                ),
-                (CallContextField::Value, call.value),
-                (CallContextField::IsStatic, (call.is_static as usize).into()),
-                (CallContextField::LastCalleeId, 0.into()),
-                (CallContextField::LastCalleeReturnDataOffset, 0.into()),
-                (CallContextField::LastCalleeReturnDataLength, 0.into()),
-                (CallContextField::IsRoot, 1.into()),
-                (CallContextField::IsCreate, 0.into()),
-                (CallContextField::CodeHash, callee_code_hash),
-            ] {
-                state.call_context_write(&mut exec_step, call.call_id, field, value);
-            }
-
-            Ok(exec_step)
+pub fn gen_associated_steps(
+    state: &mut CircuitInputStateRef,
+    execution_step: ExecState,
+) -> Result<ExecStep, Error> {
+    let fn_gen_associated_steps = match execution_step {
+        ExecState::BeginTx | ExecState::EndTx => BeginEndTx::gen_associated_steps,
+        _ => {
+            evm_unimplemented!("unsupported execution step {:?}", execution_step);
+            Dummy::gen_associated_steps
```

Comment:
I think we could just mark it as `unreachable` here:

```suggestion
            unreachable!()
```